File: BinaryTree/BinaryTree/BinaryTree.cpp
#include "BinaryTree.h"


BinaryTree::BinaryTree()
{
	this->root = nullptr;
	size = 0;
}


BinaryTree::~BinaryTree()
{
	delete root;
}

BinaryTree::BinaryTree(const BinaryTree & copyTree)
{

}

void BinaryTree::insert(int data)
{
	size++;
	std::string location = itob(size);
	BinaryTreeNode * newNode = new BinaryTreeNode(data);
	BinaryTreeNode * walker = root;
	if (size == 1) {
		root = newNode;
	}
	else
	{
		location = location.substr(1, location.size());
		for (auto& it : location) {
			if (it == '1') {
				if (walker->getRight()) {
					walker = walker->getRight();
				}
				else
				{
					newNode->setParent(walker);
					walker->setRight(newNode);
				}

			}
			else
			{
				if (walker->getLeft()) {
					walker = walker->getLeft();
				}
				else
				{
					newNode->setParent(walker);
					walker->setLeft(newNode);
				}
			}
		}
	}
}



void BinaryTree::print(BinaryTreeNode * node)
{
	if (!node)
		return;
	std::cout<<node->getData() << " " ;
	print(node->getLeft());
	print(node->getRight());

}

std::string BinaryTree::itob(int num) {
	std::string result = "";
	while (num > 0) {
		result += (num % 2 == 0) ? "0" : "1";
		num = num / 2;
	}
	std::reverse(result.begin(), result.end());
	return result;
}

BinaryTreeNode * BinaryTree::getRoot()
{
	return root;
}

File: Java Practice/Data Structure/BinaryTreeGenerics/src/BinaryTree.java



public class BinaryTree<T extends Comparable<T>> {
	private BinaryTreeNode<T> root;
	
	public BinaryTree() {
		root = null;
	}
	
	public BinaryTree(BinaryTreeNode<T> newNode ) {
		root = new BinaryTreeNode<T>(newNode);
	}
	
	public BinaryTreeNode<T>getRoot(){
		return root;
	}
	
	public void add(T data) {
		if(root == null) {
			root = new BinaryTreeNode<>(data);
		}
		else {
			add(root, data);
		}
	}
	
	public void add (BinaryTreeNode<T> node, T data) {
		int cmp = node.getData().compareTo(data);
		if(cmp > 0) {
			if(node.getLeft() == null) {
				node.setLeft(new BinaryTreeNode<>(data));
			}
			else {
				add(node.getLeft(), data);
			}
		}
		else if(cmp <= 0) {
			if(node.getRight() == null) {
				node.setRight(new BinaryTreeNode<>(data));
			}
			else {
				add(node.getRight(), data);
			}
		}
		
	}
	
	public void inOrderPrint() {
		inOrderPrint(root);
	}
	
	public void inOrderPrint(BinaryTreeNode<T> node) {
		if(node.getLeft() != null) {
			inOrderPrint(node.getLeft());
		}
		System.out.print(node.getData() + "  ");
		if(node.getRight() != null) {
			inOrderPrint(node.getRight());
		}
		
		
	}
	
}
